File: Test/python/dual_agents.py
"""
dual_agent.py  –  Use Case 2
================================
Two agents each control one game character simultaneously.
They can be teammates or enemies.

HOW CAMERA / VIEW WORKS
------------------------
- The first agent to connect (agent_01) calls SET_VIEW to claim the display.
- The second agent (agent_02) also calls SET_VIEW, but its attempt is ignored
  because the first-connected agent already owns the view.
- You can override with:
      await set_view(ws2, target_agent_id="agent_02")
  ... which hands control of the view to agent_02's character.
- In the Unity Editor/build, a dropdown in the top-right corner lets you
  manually switch the view at any time.

TYPICAL WORKFLOW
----------------
Terminal 1:  python dual_agent.py --agent agent_01
Terminal 2:  python dual_agent.py --agent agent_02

Or run both agents in the same process (this file).

Requirements:
    pip install websockets
"""

import asyncio
import argparse
import json
import time
import websockets


UNITY_WS_URL = "ws://localhost:8080/agent"


# ---------------------------------------------------------------------------
# Wall-collision / stuck detector (one instance per agent)
# ---------------------------------------------------------------------------

class StuckDetector:
    STUCK_FRAMES = 3
    POS_EPSILON  = 0.02   # metres

    def __init__(self):
        self._last_pos    = None
        self._stuck_count = 0
        self.is_moving    = False

    def on_position(self, pos: dict) -> bool:
        if not self.is_moving:
            self._stuck_count = 0
            self._last_pos = None
            return False
        cur = (pos.get("x", 0.0), pos.get("y", 0.0), pos.get("z", 0.0))
        if self._last_pos is not None:
            if all(abs(cur[i] - self._last_pos[i]) < self.POS_EPSILON for i in range(3)):
                self._stuck_count += 1
            else:
                self._stuck_count = 0
        self._last_pos = cur
        return self._stuck_count >= self.STUCK_FRAMES


_detectors: dict[str, StuckDetector] = {}


def get_detector(agent_id: str) -> StuckDetector:
    if agent_id not in _detectors:
        _detectors[agent_id] = StuckDetector()
    return _detectors[agent_id]


# ---------------------------------------------------------------------------
# Low-level helpers (same as single_agent.py)
# ---------------------------------------------------------------------------

def make_command(command_type: str, data: dict, agent_id: str) -> str:
    return json.dumps({
        "commandType": command_type,
        "data":        data,
        "agentId":     agent_id,
        "timestamp":   time.time(),
    })


async def send(ws, command_type: str, data: dict, agent_id: str):
    msg = make_command(command_type, data, agent_id)
    await ws.send(msg)
    print(f"  [{agent_id}] → {command_type} {data}")


async def move(ws, agent_id, x, z):
    d = get_detector(agent_id); d.is_moving = True; d._stuck_count = 0
    await send(ws, "MOVE", {"x": x, "z": z}, agent_id)
async def stop(ws, agent_id):
    get_detector(agent_id).is_moving = False
    await send(ws, "STOP", {}, agent_id)
async def look(ws, agent_id, p, y):      await send(ws, "LOOK",         {"pitch": p, "yaw": y},           agent_id)
async def shoot(ws, agent_id, a, d=0):   await send(ws, "SHOOT",        {"active": a, "duration": d},     agent_id)
async def reload_w(ws, agent_id):        await send(ws, "RELOAD",       {},                               agent_id)
async def switch_w(ws, agent_id, slot):  await send(ws, "SWITCH_WEAPON",{"weaponIndex": slot},            agent_id)
async def aim(ws, agent_id, a):          await send(ws, "AIM",          {"active": a},                    agent_id)
async def set_view(ws, source_id, target_id=None):
    t = target_id if target_id else source_id
    await send(ws, "SET_VIEW", {"viewTargetAgentId": t}, source_id)


# ---------------------------------------------------------------------------
# State listener
# ---------------------------------------------------------------------------

async def listen_state(ws, agent_id: str):
    async for raw in ws:
        try:
            msg = json.loads(raw)
            if msg.get("type") == "welcome":
                print(f"[{agent_id}] Connected – session {msg.get('sessionId')}")
                continue
            if "player" in msg:
                p = msg["player"]
                print(
                    f"  [{agent_id}] HP:{p.get('health',0):.0f} "
                    f"Ammo:{p.get('currentAmmo',0)} "
                    f"State:{p.get('movementState','?')}"
                )
                d = get_detector(agent_id)
                if "position" in p and d.on_position(p["position"]):
                    print(f"  [{agent_id}] Wall detected — auto-stop.")
                    d.is_moving = False
                    await send(ws, "STOP", {}, agent_id)
        except Exception:
            pass


# ---------------------------------------------------------------------------
# Per-agent scenario
# ---------------------------------------------------------------------------

async def run_agent(agent_id: str, claim_view: bool):
    """
    Connect to Unity, optionally claim the camera, then run a short scenario.
    claim_view=True  → this agent's character is shown on the Unity display.
    """
    print(f"[{agent_id}] Connecting to {UNITY_WS_URL} ...")
    async with websockets.connect(UNITY_WS_URL) as ws:
        listener = asyncio.create_task(listen_state(ws, agent_id))

        await asyncio.sleep(0.3)

        # --- View ownership ---
        if claim_view:
            print(f"\n[{agent_id}] Claiming camera view...")
            await set_view(ws, agent_id)

        # --- Scenario ---
        await asyncio.sleep(0.5)
        print(f"\n[{agent_id}] Moving forward...")
        await move(ws, agent_id, 0.0, 1.0)
        await asyncio.sleep(3.0)

        print(f"\n[{agent_id}] Looking left 45°...")
        await look(ws, agent_id, 0.0, -45.0)
        await asyncio.sleep(1.0)

        print(f"\n[{agent_id}] Shooting burst...")
        await shoot(ws, agent_id, True, 0.4)
        await asyncio.sleep(0.8)

        print(f"\n[{agent_id}] Reloading...")
        await reload_w(ws, agent_id)
        await asyncio.sleep(2.5)

        print(f"\n[{agent_id}] Stop.")
        await stop(ws, agent_id)

        # --- Optionally hand view to other agent ---
        if claim_view:
            other = "agent_02" if agent_id == "agent_01" else "agent_01"
            print(f"\n[{agent_id}] Handing view to {other}...")
            await set_view(ws, agent_id, target_id=other)
            await asyncio.sleep(3.0)

        listener.cancel()
        print(f"[{agent_id}] Done.")


# ---------------------------------------------------------------------------
# Entry points
# ---------------------------------------------------------------------------

async def run_both_in_one_process():
    """Run agent_01 and agent_02 concurrently in the same script."""
    print("=== Dual Agent Demo (both in one process) ===\n")
    await asyncio.gather(
        run_agent("agent_01", claim_view=True),
        run_agent("agent_02", claim_view=False),
    )


async def run_single(agent_id: str):
    """Run one agent (meant to be launched from a separate terminal)."""
    claim = (agent_id == "agent_01")
    await run_agent(agent_id, claim_view=claim)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Dual-agent WebSocket client")
    parser.add_argument(
        "--agent",
        default=None,
        help="Agent ID to run ('agent_01' or 'agent_02'). "
             "Omit to run both concurrently in one process.",
    )
    args = parser.parse_args()

    if args.agent:
        asyncio.run(run_single(args.agent))
    else:
        asyncio.run(run_both_in_one_process())

File: Test/javascript/dual_agents.js
/**
 * dual_agents.js  –  Use Case 2
 * ================================
 * Two agents each control one game character simultaneously.
 * They can be teammates or enemies.
 *
 * HOW CAMERA / VIEW WORKS
 * ------------------------
 * - The first agent to connect (agent_01) calls SET_VIEW to claim the display.
 * - agent_02 can take over by calling SET_VIEW with its own agentId.
 * - Either agent can point the view at the other with viewTargetAgentId.
 * - In Unity Editor / Development builds a dropdown in the top-right corner
 *   lets you manually switch the view at any time.
 *
 * TYPICAL WORKFLOW
 * ----------------
 * Terminal 1:  node dual_agents.js --agent agent_01
 * Terminal 2:  node dual_agents.js --agent agent_02
 *
 * Or run both agents in the same process (default):
 *   node dual_agents.js
 *
 * Requirements:
 *   npm install ws        (already in package.json)
 */

'use strict';

const WebSocket = require('ws');

const UNITY_WS_URL = 'ws://localhost:8080/agent';

// ---------------------------------------------------------------------------
// Low-level helpers
// ---------------------------------------------------------------------------

function makeCommand(commandType, data = {}, agentId) {
  return JSON.stringify({
    commandType,
    data,
    agentId,
    timestamp: Date.now() / 1000,
  });
}

function send(ws, commandType, data = {}, agentId) {
  ws.send(makeCommand(commandType, data, agentId));
  console.log(`  [${agentId}] → ${commandType}`, data);
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// ---------------------------------------------------------------------------
// Wall-collision / stuck detector (one instance per agent)
// ---------------------------------------------------------------------------

const STUCK_FRAMES = 3;
const POS_EPSILON  = 0.02;  // metres

const detectors = {};

function getDetector(agentId) {
  if (!detectors[agentId]) {
    detectors[agentId] = {
      isMoving:   false,
      lastPos:    null,
      stuckCount: 0,
      onPosition(pos) {
        if (!this.isMoving) { this.stuckCount = 0; this.lastPos = null; return false; }
        if (this.lastPos) {
          const same =
            Math.abs(pos.x - this.lastPos.x) < POS_EPSILON &&
            Math.abs(pos.y - this.lastPos.y) < POS_EPSILON &&
            Math.abs(pos.z - this.lastPos.z) < POS_EPSILON;
          this.stuckCount = same ? this.stuckCount + 1 : 0;
        }
        this.lastPos = { ...pos };
        return this.stuckCount >= STUCK_FRAMES;
      },
    };
  }
  return detectors[agentId];
}

// ---------------------------------------------------------------------------
// High-level actions (all require explicit agentId)
// ---------------------------------------------------------------------------

const move = (ws, id, x, z) => {
  const d = getDetector(id); d.isMoving = true; d.stuckCount = 0;
  send(ws, 'MOVE', { x, z }, id);
};
const stop = (ws, id) => {
  getDetector(id).isMoving = false;
  send(ws, 'STOP', {}, id);
};
const look         = (ws, id, pitch, yaw) => send(ws, 'LOOK',         { pitch, yaw },          id);
const shoot        = (ws, id, active, d=0)=> send(ws, 'SHOOT',        { active, duration: d }, id);
const reloadWeapon = (ws, id)             => send(ws, 'RELOAD',       {},                      id);
const switchWeapon = (ws, id, slot)       => send(ws, 'SWITCH_WEAPON',{ weaponIndex: slot },   id);
const aim          = (ws, id, active)     => send(ws, 'AIM',          { active },              id);
const setView      = (ws, sourceId, targetId) =>
  send(ws, 'SET_VIEW', { viewTargetAgentId: targetId || sourceId }, sourceId);

// ---------------------------------------------------------------------------
// Game-state listener
// ---------------------------------------------------------------------------

function makeListener(agentId, ws) {
  return function onMessage(raw) {
    try {
      const msg = JSON.parse(raw);
      if (msg.type === 'welcome') {
        console.log(`[${agentId}] Connected – session ${msg.sessionId}`);
        return;
      }
      if (msg.player) {
        const p = msg.player;
        console.log(
          `  [${agentId}] HP:${p.health} ` +
          `Ammo:${p.currentAmmo} ` +
          `State:${p.movementState}`
        );
        const d = getDetector(agentId);
        if (p.position && d.onPosition(p.position)) {
          console.log(`  [${agentId}] Wall detected — auto-stop.`);
          d.isMoving = false;
          send(ws, 'STOP', {}, agentId);
        }
      }
    } catch (_) { /* ignore */ }
  };
}

// ---------------------------------------------------------------------------
// Per-agent scenario
// ---------------------------------------------------------------------------

async function connectAgent(agentId) {
  console.log(`[${agentId}] Connecting to ${UNITY_WS_URL} ...`);
  const ws = new WebSocket(UNITY_WS_URL);
  await new Promise((resolve, reject) => {
    ws.once('open',  resolve);
    ws.once('error', reject);
  });
  ws.on('message', makeListener(agentId, ws));
  ws.on('close', (code) => console.log(`[${agentId}] Closed (${code})`));
  return ws;
}

async function runAgent(agentId, claimView) {
  const ws = await connectAgent(agentId);
  const other = agentId === 'agent_01' ? 'agent_02' : 'agent_01';

  await sleep(300);

  // --- View ownership ---
  if (claimView) {
    console.log(`\n[${agentId}] Claiming camera view...`);
    setView(ws, agentId);
  }

  await sleep(500);

  console.log(`\n[${agentId}] Moving forward...`);
  move(ws, agentId, 0, 1);
  await sleep(3000);

  console.log(`\n[${agentId}] Looking left 45°...`);
  look(ws, agentId, 0, -45);
  await sleep(1000);

  console.log(`\n[${agentId}] Shooting burst...`);
  shoot(ws, agentId, true, 0.4);
  await sleep(800);

  console.log(`\n[${agentId}] Reloading...`);
  reloadWeapon(ws, agentId);
  await sleep(2500);

  console.log(`\n[${agentId}] Stop.`);
  stop(ws, agentId);

  // --- Optionally hand view to the other agent ---
  if (claimView) {
    console.log(`\n[${agentId}] Handing view to ${other}...`);
    setView(ws, agentId, other);
    await sleep(3000);
  }

  ws.close();
  console.log(`[${agentId}] Done.`);
}

// ---------------------------------------------------------------------------
// Entry points
// ---------------------------------------------------------------------------

async function runBothInOneProcess() {
  console.log('=== Dual Agent Demo (both in one process) ===\n');
  await Promise.all([
    runAgent('agent_01', /* claimView */ true),
    runAgent('agent_02', /* claimView */ false),
  ]);
}

async function runSingle(agentId) {
  await runAgent(agentId, agentId === 'agent_01');
}

// Parse optional --agent flag
const args    = process.argv.slice(2);
const agentIdx = args.indexOf('--agent');
const agentArg = agentIdx !== -1 ? args[agentIdx + 1] : null;

if (agentArg) {
  runSingle(agentArg).catch(err => {
    console.error('Fatal error:', err.message);
    process.exit(1);
  });
} else {
  runBothInOneProcess().catch(err => {
    console.error('Fatal error:', err.message);
    process.exit(1);
  });
}
